Use pkg/forwarder for submitting process-agent payloads

kevinconaway opened the pull request:
### What does this PR do?

This commit modifies the process-agent to use the pkg/forwarder facility for sending HTTP payloads.  This will ensure that payloads are retried until they are successful.

One difference is that this will submit the API key in a different header than we are currently using.  The intake API will be modified to read this header.

We're also sending a new header, X-DD-Agent-Timestamp that indicates the time that we attempted to send the payload

### Motivation

What inspired you to submit this pull request?

### Additional Notes

QA:

To QA, we need to make sure that process, container, pod and connection messages are all correctly sent to intake with. and without an HTTP proxy

Additionally, we want make sure that real time mode is toggled on and off appropriately

kevinconaway committed:
Use pkg/forwarder for submitting process-agent payloads

This commit modifies the process-agent to use the pkg/forwarder facility for sending HTTP payloads.  This will ensure that payloads are retried until they are successful.

One difference is that this will submit the API key in a different header than we are currently using.  The intake API will be modified to read this header.

We're also sending a new header, X-DD-Agent-Timestamp that indicates the time that we attempted to send the payload

kevinconaway commented:
One other point to note here is that this will now use the main agent proxy configuration instead of the process agent proxy configuration.  My assumption was that this would not be a problem because if the process agent is proxied than the main agent would be as well, with the same configuration.

mfpierre commented on pkg/clusteragent/orchestrator/controller.go:
```diff
@@ -134,13 +151,41 @@ func (o *Controller) processPods() {
+		}
+
+		payloads := forwarder.Payloads{&body}
+		responses, err := o.forwarder.SubmitPodChecks(payloads, extraHeaders)
```
```suggestion
		_, err := o.forwarder.SubmitPodChecks(payloads, extraHeaders)
```
we can ignore responses here and remove the range below

kevinconaway replied:
I had left that in as a precautionary measure in case the resulting channel was changed to be unbuffered or of a different size than it currently is.  As of now, the channel is sized to the number of output transactions so none of the worker threads will be blocked writing to it but if the channel size was changed, the worker threads could be blocked if nothing is consuming from the channel.

Thoughts?

mfpierre replied:
sounds good to me, let's keep it 👍

mfpierre commented on cmd/process-agent/collector.go:
```diff
@@ -134,17 +147,47 @@ func (l *Collector) run(exit chan bool) {
+					case checks.Connections.Name():
+						responses, err = l.forwarder.SubmitConnectionChecks(payloads, extraHeaders)
+					case checks.Pod.Name():
+						responses, err = l.forwarder.SubmitPodChecks(payloads, extraHeaders)
```
```suggestion
						responses, err = l.podForwarder.SubmitPodChecks(payloads, extraHeaders)
```

kevinconaway committed:
Add more documentation around why we're consuming the responses channel

kevinconaway committed:
Use a separate mock endpoint for orchestrator payloads

This will ensure that we test that orchestrator payloads are sent to the orchestrator enpdoint

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -67,7 +69,8 @@ func NewCollector(cfg *config.AgentConfig) (Collector, error) {
 		cfg:           cfg,
 		groupID:       rand.Int31(),
-		apiClient:     api.NewClient(http.Client{Timeout: HTTPTimeout, Transport: cfg.Transport}, ReqCtxTimeout),
+		forwarder:     forwarder.NewDefaultForwarder(keysPerDomains(cfg.APIEndpoints)),
```
This is a convenient way of setting things up :)

sunhay replied:
Does the `DefaultForwarder` setup it's own `Transport` using the config? We'll want to make sure it does so that proxy settings continue to work

kevinconaway replied:
Yes.  Per my comment above, it uses the proxy settings from the main agent config though.

My assumption was that if proxying was configured on the process agent then the same configuration would be applied to the main agent.

sunhay replied:
It is - this is great :)

leeavital commented:
Please add QA steps in the PR description we can use during the release process

leeavital commented on cmd/process-agent/collector_api_test.go:
```diff
@@ -19,56 +22,222 @@ import (
+
+		assert.Equal(t, "1", req.headers.Get("X-DD-Agent-Attempts"))
+
+		select {
```
could we have runCollectorTest read until `payloads` is closed and then flush to get rid of the need to sleep here?

kevinconaway replied:
The sleep isn't ideal but I'm not aware a better way to test that something _doesn't_ happen, when accounting for an asynchronous i/o call like this.

We don't have access to the `payloads` channel here unless I'm misunderstanding what you're suggesting.

mfpierre commented on pkg/clusteragent/orchestrator/controller.go:
```diff
@@ -8,11 +8,16 @@
 
+	model "github.com/DataDog/agent-payload/process"
+
+	"github.com/DataDog/datadog-agent/pkg/forwarder"
```
nit: group this up with import below

kevinconaway replied:
Hmm, this is how go imports formats them for me.   Does it do something different for you?

mfpierre replied:
in the agent we usually have 3 groups for imports
- native
- project
- 3rd party

IIRC, if you group them up with the ones below they should stick to the group and be ordered alphabetically

mfpierre reviewed (dismissed):
Looking good on the container-app side, nice rework 🙌

shang-wang commented on pkg/forwarder/forwarder.go:
```diff
@@ -346,3 +383,87 @@ func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) e
+
+	results := make(chan Response, len(transactions))
+
+	var wg sync.WaitGroup
```
What's async here that a wg is needed?

kevinconaway replied:
Submitting the "http transaction" to the workers.  We will create txn per (domain,  apiKey) combination and each will be submitted. separately.

When all. the requests complete, the results channel will be closed (via a separate goroutine waiting on the wait group)

shang-wang replied:
Ah got it.

shang-wang commented on pkg/clusteragent/orchestrator/controller.go:
```diff
@@ -107,13 +118,19 @@ func (o *Controller) Run(stopCh <-chan struct{}) {
 	defer log.Infof("Stopping orchestrator controller")
 
+	if err := o.forwarder.Start(); err != nil {
+		log.Errorf("error starting pod forwarder: %s", err)
```
If forwarder didn't start we don't need to return?

sunhay commented on cmd/process-agent/collector.go:
```diff
@@ -118,6 +121,16 @@ func (l *Collector) run(exit chan bool) {
+		return
+	}
+
+	if err := l.podForwarder.Start(); err != nil {
```
Should we only do this if the `orchestrator` check is enabled?

kevinconaway replied:
I guess we could, we would have to surround this and the stop below with an if check.  Is it worth it to avoid having idle goroutines sitting around?

sunhay commented on pkg/forwarder/transaction.go:
```diff
@@ -210,7 +210,7 @@ func (t *HTTPTransaction) internalProcess(ctx context.Context, client *http.Clie
 		return 0, nil, fmt.Errorf("error while sending transaction, rescheduling it: %s", httputils.SanitizeURL(err.Error()))
 	}
-	defer resp.Body.Close()
+	defer func() { _ = resp.Body.Close() }()
```
Nice catch - this means there was a leak beforehand, no?

kevinconaway replied:
Nope, this is just removing an IDE warning about ignoring an error response.

prognant commented on cmd/process-agent/collector.go:
```diff
@@ -258,3 +289,63 @@ func getContainerCount(mb model.MessageBody) int {
+			continue
+		}
+
+		if response.StatusCode >= 300 {
+			log.Errorf("Invalid response from %s: %d -> %s", response.Domain, response.StatusCode, response.Err)
+			continue
+		}
```
Are we 100% sure that 3xx / redirect code should be considered as invalid ?

kevinconaway replied:
Yes, this would normally be handled inside the `net/http`.  Anything that makes its way back to the caller that isn't 2xx is invalid.

hush-hush commented on pkg/forwarder/transaction.go:
```diff
@@ -210,7 +210,7 @@ func (t *HTTPTransaction) internalProcess(ctx context.Context, client *http.Clie
 		return 0, nil, fmt.Errorf("error while sending transaction, rescheduling it: %s", httputils.SanitizeURL(err.Error()))
 	}
-	defer resp.Body.Close()
+	defer func() { _ = resp.Body.Close() }()
```
Why this change ?

kevinconaway replied:
This removes a warning about ignoring an error being returned (`resp.Body.Close()` returns an error)

hush-hush commented on pkg/forwarder/forwarder.go:
```diff
@@ -346,3 +383,87 @@ func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) e
+		}
+	}
+
+	go func() {
+		wg.Wait()
+		close(results)
+	}()
```
This will create a goroutine by transaction currently being queued. We might drop a transaction without calling the completion handler for a number of reasons ([retry queue is fulled](https://github.com/DataDog/datadog-agent/blob/master/pkg/forwarder/domain_forwarder.go#L108), the domain was black listed, ...).

I don't think we should rely on it being closed or add a dropped handler and call it from everywhere. I'm less in favor of the second as it's so easy to forget to call it in future development leading to a major leak in the agent.

kevinconaway replied:
Good point, I will rewrite this so that there is not a reliance on the channel being closed.

kevinconaway committed:
Return early if the orchestrator forwarder does not start

kevinconaway committed:
Fix import formatting per agent team style guidelines

kevinconaway committed:
Log an error if we encounter an unknown payload type to submit

kevinconaway committed:
Ensure that the process agent stops if the forwarders cannot be started

kevinconaway committed:
Merge branch 'master' of github.com:DataDog/datadog-agent into kevinconaway/process-pkg-forwarder

 Conflicts:
	pkg/process/checks/container_nolinux.go
	pkg/process/checks/container_rt_nolinux.go
	pkg/process/checks/pod_nolinux.go

kevinconaway committed:
Ensure that response channel is closed and not left dangling

This commit ensures that the response channel is always closed for the caller by reading from a separate internal results channel and waiting for the expected number of responses.  If the number of expected responses does not arrive within a a timeout period, the results channel is closed and the goroutine gives up on waiting

kevinconaway committed:
Merge branches 'kevinconaway/process-pkg-forwarder' and 'master' of github.com:DataDog/datadog-agent into kevinconaway/process-pkg-forwarder

hush-hush commented on pkg/forwarder/forwarder.go:
```diff
@@ -346,3 +387,98 @@ func (f *DefaultForwarder) SubmitV1Intake(payload Payloads, extra http.Header) e
+				results <- r
+				receivedResponses++
+				if receivedResponses == expectedResponses {
+					close(results)
```
Should we not return where ? Seems like we're going to run the loop until the timeout otherwise and double close a channel (which will panic).

kevinconaway committed:
Ensure that the goroutine reading responses exits after receiving all expected messages

ahmed-mez reviewed (approved):
Reviewed only the changes in `pkg/clusteragent` owned by the team, LGTM 👍
